src/v/cloud_storage/remote_partition.cc

```diff
@@ -0,0 +1,641 @@
+    while (readahead--) {
+        pi++;
+        if (pi != _segments.end()) {
+            std::visit(
+              visit_hydrate{
+                .state = pi->second, .part = this, .offset_key = pi->first},
+              pi->second);
+        } else {
+            break;
+        }
+    }
+}
+
+void remote_partition::gc_stale_materialized_segments() {
+    vlog(
+      _ctxlog.debug,
+      "collecting stale materialized segments, {} segments materialized, {} "
+      "segments total",
+      _materialized.size(),
+      _segments.size());
+    auto now = ss::lowres_clock::now();
+    std::vector<model::offset> offsets;
+    for (auto& st : _materialized) {
+        if (now - st.atime > stm_max_idle_time && st.segment.owned()) {
+            vlog(
+              _ctxlog.debug,
+              "reader for segment with base offset {} is stale",
+              st.offset_key);
+            // this will delete and unlink the object from
+            // _materialized collection
+            offsets.push_back(st.offset_key);
+        }
+    }
+    vlog(_ctxlog.debug, "found {} eviction candidates ", offsets.size());
+    for (auto o : offsets) {
+        vlog(_ctxlog.debug, "about to offload segment {}", o);
+        auto it = _segments.find(o);
+        vassert(it != _segments.end(), "Can't find offset {}", o);
+        auto tmp = std::visit(
+          [this](auto&& st) { return st->offload(this); }, _segments[o]);
+        _segments[o] = std::move(tmp);
+    }
+}
+
+model::offset remote_partition::first_uploaded_offset() {
+    if (_manifest.size() == 0) {
+        return model::offset(0);
+    }
+    try {
+        if (_first_uploaded_offset) {
+            return *_first_uploaded_offset;
+        }
+        model::offset starting_offset = model::offset::max();
+        for (const auto& m : _manifest) {
+            starting_offset = std::min(
+              starting_offset, get_kafka_base_offset(m.second));
+            vlog(
+              _ctxlog.trace,
+              "remote partition first_uploaded_offset .. {}",
+              starting_offset);
+        }
+        _first_uploaded_offset = starting_offset;
+        vlog(
+          _ctxlog.debug,
+          "remote partition first_uploaded_offset set to {}",
+          starting_offset);
+        return starting_offset;
+    } catch (...) {
+        vlog(
+          _ctxlog.error,
+          "remote partition first_uploaded_offset error {}",
+          std::current_exception());
+
+        throw;
+    }
+}
+
+model::ntp remote_partition::get_ntp() const { return _manifest.get_ntp(); }
+
+ss::future<> remote_partition::stop() {
+    vlog(_ctxlog.debug, "remote partition stop {} segments", _segments.size());
+    _stm_timer.cancel();
+    _cvar.broken();
+
+    for (auto& [offset, seg] : _segments) {
+        vlog(_ctxlog.debug, "remote partition stop {}", offset);
+        co_await std::visit([](auto&& st) { return st->stop(); }, seg);
+    }
+
+    co_await _gate.close();
```

Comment:
do you need to close the gate before waiting on the segments to stop and then finally wait on all fibers to leave the gate before returning?

Reply:
that makes sense, fixed